From: Sipwise Development Team <support@sipwise.com>
Date: Thu, 26 Mar 2020 10:06:45 +0100
Subject: no_INSTALL_file

---
 src/Makefile | 2 --
 1 file changed, 2 deletions(-)

diff --git a/src/Makefile b/src/Makefile
index 4ba6625..edad3d2 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -859,8 +859,6 @@ install-modules-all: install-every-module install-every-module-doc
 
 
 install-doc: $(doc_prefix)/$(doc_dir) install-every-module-doc
-	$(INSTALL_TOUCH) $(doc_prefix)/$(doc_dir)/INSTALL
-	$(INSTALL_DOC) ../INSTALL $(doc_prefix)/$(doc_dir)
 	$(INSTALL_TOUCH) $(doc_prefix)/$(doc_dir)/README
 	$(INSTALL_DOC) ../README $(doc_prefix)/$(doc_dir)
 
