From: Sipwise Development Team <support@sipwise.com>
Date: Wed, 17 Feb 2021 11:46:21 +0100
Subject: cfgt_debug

---
 src/modules/cfgt/cfgt_int.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/modules/cfgt/cfgt_int.c b/src/modules/cfgt/cfgt_int.c
index e2f402f..d60cc61 100644
--- a/src/modules/cfgt/cfgt_int.c
+++ b/src/modules/cfgt/cfgt_int.c
@@ -21,6 +21,8 @@
  */
 #include <stdio.h>
 #include <sys/stat.h>
+#include <string.h>
+#include <errno.h>
 
 #include "../../core/events.h"
 #include "../../core/strutils.h"
@@ -349,7 +351,7 @@ void cfgt_save_node(cfgt_node_p node)
 			return;
 		}
 		if(fputs(dest.s, fp) < 0) {
-			LM_ERR("failed writing to file\n");
+			LM_ERR("failed writing to file: %s\n", strerror(errno));
 		}
 		fclose(fp);
 		node->jdoc.free_fn(dest.s);
