--- a/src/modules/db_redis/redis_table.c
+++ b/src/modules/db_redis/redis_table.c
@@ -589,6 +589,10 @@
                 column_name.s = start;
                 column_name.len = p - start;
                 start = ++p;
+
+                if (!column_name.len)
+                    break;
+
                 /*
                 LM_DBG("found column name '%.*s' in type '%.*s' for table '%.*s'\n",
                         column_name.len, column_name.s,
--- a/src/modules/db_redis/redis_dbase.c
+++ b/src/modules/db_redis/redis_dbase.c
@@ -383,6 +383,22 @@
         }
     }
 
+    // for value-less master keys
+    if (!key_name->len) {
+        // <version>:<table_name>:<type>
+        len = table->version_code.len + table_name->len + 1 + type_name->len + 1;
+        key_name->s = (char*)pkg_malloc(len);
+        if (!key_name->s) {
+            LM_ERR("Failed to allocate key memory\n");
+            goto err;
+        }
+        snprintf(key_name->s, len, "%.*s%.*s:%.*s",
+                table->version_code.len, table->version_code.s,
+                table_name->len, table_name->s,
+                type_name->len, type_name->s);
+        key_name->len = len-1;
+    }
+
     return 0;
 
 err:
@@ -2052,9 +2068,9 @@
                     type_key->key.len, type_key->key.s);
             char *prefix = ser_memmem(type_key->key.s, "::", type_key->key.len, 2);
             if (!prefix || prefix == type_key->key.s) {
-                LM_ERR("Invalid key without :: '%.*s'\n",
+                LM_DBG("Invalid key without :: '%.*s'\n",
                         type_key->key.len, type_key->key.s);
-                goto error;
+                continue;
             }
             for (new_type_key = new_type_keys; new_type_key; new_type_key = new_type_key->next) {
                 // compare prefix to see if this is the same key
