From: Victor Seva <vseva@sipwise.com>
Date: Tue, 13 Dec 2022 11:38:00 +0100
Subject: dialog: dlg_get_var changes in API

---
 src/modules/dialog/dialog.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/modules/dialog/dialog.c b/src/modules/dialog/dialog.c
index 6545444..25f3ca1 100644
--- a/src/modules/dialog/dialog.c
+++ b/src/modules/dialog/dialog.c
@@ -1536,7 +1536,7 @@ static str *ki_dlg_get_var_helper(sip_msg_t *msg, str *sc, str *sf, str *st, str
 	dlg = get_dlg(sc, sf, st, &dir);
 	if(dlg==NULL)
 		return val;
-	val = get_dlg_variable(dlg, key);
+	val = get_dlg_varref(dlg, key);
 	dlg_release(dlg);
 	return val;
 }
