From: Sipwise Development Team <support@sipwise.com>
Date: Wed, 5 Aug 2026 10:44:17 +0200
Subject: dialog_w_dlg_get_var_no_error_msg

---
 src/modules/dialog/dialog.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/modules/dialog/dialog.c b/src/modules/dialog/dialog.c
index 4b681b2..538e383 100644
--- a/src/modules/dialog/dialog.c
+++ b/src/modules/dialog/dialog.c
@@ -1979,7 +1979,7 @@ static int w_dlg_get_var(
 		pv_get_null(msg, NULL, &dst_val);
 	}
 	if(pv_set_spec_value(msg, dst_pv, 0, &dst_val) != 0) {
-		LM_ERR("unable to set value to dst_pv\n");
+		LM_INFO("unable to set value to dst_pv\n");
 		if(val)
 			goto error;
 		else
@@ -1990,7 +1990,7 @@ static int w_dlg_get_var(
 error:
 	pv_get_null(msg, NULL, &dst_val);
 	if(pv_set_spec_value(msg, dst_pv, 0, &dst_val) != 0) {
-		LM_ERR("unable to set null value to dst_pv\n");
+		LM_INFO("unable to set null value to dst_pv\n");
 	}
 	return -1;
 }
