--- a/src/modules/dialog/dialog.c
+++ b/src/modules/dialog/dialog.c
@@ -2159,7 +2159,7 @@ static int w_dlg_get_var(
 	}
 	if(pv_set_spec_value(msg, dst_pv, 0, &dst_val) != 0) {
 		if (dst_pv && dst_pv->type != PVT_OTHER)
-			LM_ERR("unable to set value to dst_pv\n");
+			LM_INFO("unable to set value to dst_pv\n");
 		if(val)
 			goto error;
 		else
@@ -2171,7 +2171,7 @@ error:
 	pv_get_null(msg, NULL, &dst_val);
 	if(pv_set_spec_value(msg, dst_pv, 0, &dst_val) != 0) {
 		if (dst_pv && dst_pv->type != PVT_OTHER)
-			LM_ERR("unable to set null value to dst_pv\n");
+			LM_INFO("unable to set null value to dst_pv\n");
 	}
 	return -1;
 }
