From: Sipwise Development Team <support@sipwise.com>
Date: Thu, 26 Mar 2020 10:06:45 +0100
Subject: do-not-install-default-config

---
 src/Makefile          | 2 +-
 utils/kamctl/Makefile | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/Makefile b/src/Makefile
index 8de3eb2..63644c8 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -650,7 +650,7 @@ sunpkg:
 
 .PHONY: install
 install: mk_params="compile_for_install=yes"
-install: install-bin install-every-module install-cfg \
+install: install-bin install-every-module \
 	install-doc install-man install-utils install-share
 
 .PHONY: dbinstall
diff --git a/utils/kamctl/Makefile b/utils/kamctl/Makefile
index 9736cf2..4249d08 100644
--- a/utils/kamctl/Makefile
+++ b/utils/kamctl/Makefile
@@ -13,7 +13,7 @@ all:
 
 install-if-newer: install
 
-install: install-cfg install-bin install-man install-modules
+install: install-bin install-man install-modules
 
 install-cfg: $(cfg_prefix)/$(cfg_dir)
 		# kamctl config
