--- a/src/Makefile
+++ b/src/Makefile
@@ -652,7 +652,7 @@ sunpkg:
 
 .PHONY: install
 install: mk_params="compile_for_install=yes"
-install: install-bin install-every-module install-cfg \
+install: install-bin install-every-module \
 	install-doc install-man install-utils install-share
 
 .PHONY: dbinstall
--- a/utils/kamctl/Makefile
+++ b/utils/kamctl/Makefile
@@ -13,7 +13,7 @@ all:
 
 install-if-newer: install
 
-install: install-cfg install-bin install-man install-modules
+install: install-bin install-man install-modules
 
 install-cfg: $(cfg_prefix)/$(cfg_dir)
 		# kamctl config
