Index: kamailio/src/modules/db_redis/redis_dbase.c
===================================================================
--- kamailio.orig/src/modules/db_redis/redis_dbase.c
+++ kamailio/src/modules/db_redis/redis_dbase.c
@@ -236,6 +236,7 @@ static int db_redis_build_entry_manual_k
     for (key = table->entry_keys; key; key = key->next) {
         int subkey_found = 0;
         int i;
+        *manual_key_count = 0;
         LM_DBG("checking for existence of entry key '%.*s' in query to get manual key\n",
                 key->key.len, key->key.s);
         for (i = 0; i < _n; ++i) {
@@ -1080,7 +1081,12 @@ static int db_redis_perform_query(const 
     RES_COL_N(*_r) = _nc;
 
     if (!(*keys_count) && do_table_scan) {
-        LM_DBG("performing full table scan\n");
+        LM_WARN("performing full table scan on table '%.*s' while performing query\n",
+                CON_TABLE(_h)->len, CON_TABLE(_h)->s);
+        for(i = 0; i < _n; ++i) {
+            LM_WARN("  scan key %d is '%.*s'\n",
+                    i, _k[i]->len, _k[i]->s);
+        }
         if (db_redis_scan_query_keys(con, CON_TABLE(_h), _k, _n,
                     keys, keys_count,
                     manual_keys, manual_keys_count) != 0) {
@@ -1232,10 +1238,10 @@ error:
 
 static int db_redis_perform_delete(const db1_con_t* _h, km_redis_con_t *con, const db_key_t* _k,
         const db_val_t* _v, const db_op_t *_op, const int _n,
-        redis_key_t *keys, int keys_count,
-        int *manual_keys, int manual_keys_count, int do_table_scan) {
+        redis_key_t **keys, int *keys_count,
+        int **manual_keys, int *manual_keys_count, int do_table_scan) {
 
-    int j = 0;
+    int i = 0, j = 0;
     redis_key_t *k = NULL;
     int type_keys_count = 0;
     int all_type_keys_count = 0;
@@ -1249,11 +1255,16 @@ static int db_redis_perform_delete(const
     db_key_t *db_keys = NULL;
     redis_key_t *type_key;
 
-    if (!keys_count && do_table_scan) {
-        LM_DBG("performing full table scan\n");
+    if (!*keys_count && do_table_scan) {
+        LM_WARN("performing full table scan on table '%.*s' while performing delete\n",
+                CON_TABLE(_h)->len, CON_TABLE(_h)->s);
+        for(i = 0; i < _n; ++i) {
+            LM_WARN("  scan key %d is '%.*s'\n",
+                    i, _k[i]->len, _k[i]->s);
+        }
         if (db_redis_scan_query_keys(con, CON_TABLE(_h), _k, _n,
-                    &keys, &keys_count,
-                    &manual_keys, &manual_keys_count) != 0) {
+                    keys, keys_count,
+                    manual_keys, manual_keys_count) != 0) {
             LM_ERR("failed to scan query keys\n");
             goto error;
         }
@@ -1270,7 +1281,7 @@ static int db_redis_perform_delete(const
     }
 
     LM_DBG("delete all keys\n");
-    for (k = keys; k; k = k->next) {
+    for (k = *keys; k; k = k->next) {
         redis_key_t *all_type_key;
         str *key = &k->key;
         redis_key_t *tmp = NULL;
@@ -1292,6 +1303,7 @@ static int db_redis_perform_delete(const
         if (reply->integer == 0) {
             LM_DBG("key does not exist in redis, skip deleting\n");
             db_redis_free_reply(&reply);
+            db_redis_key_free(&query_v);
             continue;
         }
         db_redis_free_reply(&reply);
@@ -1305,8 +1317,8 @@ static int db_redis_perform_delete(const
         }
 
         // add all manual keys to query
-        for (j = 0; j < manual_keys_count; ++j) {
-            int idx = manual_keys[j];
+        for (j = 0; j < *manual_keys_count; ++j) {
+            int idx = (*manual_keys)[j];
             str *col = _k[idx];
 
             if (db_redis_key_add_str(&query_v, col) != 0) {
@@ -1331,8 +1343,8 @@ static int db_redis_perform_delete(const
         // manually filter non-matching replies
         row_match = 1;
         for (col = 0; col < reply->elements; ++col) {
-            if (col < manual_keys_count) {
-                int idx = manual_keys[col];
+            if (col < *manual_keys_count) {
+                int idx = (*manual_keys)[col];
                 db_key_t k = _k[idx];
                 db_val_t v = _v[idx];
                 db_op_t o = _op[idx];
@@ -1370,7 +1382,7 @@ static int db_redis_perform_delete(const
         for (j = 0, all_type_key = all_type_keys; all_type_key; ++j, all_type_key = all_type_key->next) {
             db_val_t *v = &(db_vals[j]);
             str *key = &all_type_key->key;
-            char *value = reply->element[manual_keys_count + j]->str;
+            char *value = reply->element[*manual_keys_count + j]->str;
             int coltype = db_redis_schema_get_column_type(con, CON_TABLE(_h), key);
             if (value == NULL) {
                 VAL_NULL(v) = 1;
@@ -1422,13 +1434,12 @@ static int db_redis_perform_delete(const
             db_redis_check_reply(con, reply, error);
             db_redis_free_reply(&reply);
         }
-
-        //db_redis_key_free(&type_keys);
         LM_DBG("done with loop '%.*s'\n", k->key.len, k->key.s);
     }
     db_redis_key_free(&type_keys);
     db_redis_key_free(&all_type_keys);
     db_redis_key_free(&query_v);
+    // TODO: fooooo free keys and manual keys?
 
     return 0;
 
@@ -1461,7 +1472,12 @@ static int db_redis_perform_update(const
     size_t col;
 
     if (!(*keys_count) && do_table_scan) {
-        LM_DBG("performing full table scan\n");
+        LM_WARN("performing full table scan on table '%.*s' while performing update\n",
+                CON_TABLE(_h)->len, CON_TABLE(_h)->s);
+        for(i = 0; i < _n; ++i) {
+            LM_WARN("  scan key %d is '%.*s'\n",
+                    i, _k[i]->len, _k[i]->s);
+        }
         if (db_redis_scan_query_keys(con, CON_TABLE(_h), _k, _n,
                     keys, keys_count,
                     manual_keys, manual_keys_count) != 0) {
@@ -2026,10 +2042,12 @@ int db_redis_delete(const db1_con_t* _h,
         keys_count = 0;
     }
 
+    /* TODO: just for troubleshooting, remove me!
     if (db_redis_perform_delete(_h, con, _k, _v, query_ops, _n,
-        keys, keys_count, manual_keys, manual_keys_count, do_table_scan) != 0) {
+        &keys, &keys_count, &manual_keys, &manual_keys_count, do_table_scan) != 0) {
         goto error;
     }
+    */
 
     LM_DBG("done performing delete\n");
 
