--- a/modules/rtpproxy-ng/rtpproxy_funcs.c
+++ b/modules/rtpproxy-ng/rtpproxy_funcs.c
@@ -190,7 +190,7 @@
 		goto error;
 	}
 
-	body->len = get_content_length(msg);
+	body->len = msg->len -(int)(body->s-msg->buf);
 	if (body->len==0) {
 		LM_ERR("message body has length zero\n");
 		goto error;
@@ -271,7 +271,7 @@
 	return -1;
 
 done:
-	/*LM_DBG("DEBUG:extract_body:=|%.*s|\n",body->len,body->s);*/
+	LM_DBG("DEBUG:extract_body:=|%.*s|\n",body->len,body->s);
 	return 1;
 }
 
