From: Sipwise Development Team <support@sipwise.com>
Date: Mon, 2 Oct 2023 13:16:46 +0200
Subject: lcr_stats_flags_as_id

---
 src/modules/lcr/lcr_rpc.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/modules/lcr/lcr_rpc.c b/src/modules/lcr/lcr_rpc.c
index fb38912..00a7a9e 100644
--- a/src/modules/lcr/lcr_rpc.c
+++ b/src/modules/lcr/lcr_rpc.c
@@ -109,7 +109,8 @@ static void print_gw_stat(rpc_t *rpc, void *st, struct gw_info *gw)
 {
 	str gw_name;
 
-	rpc->struct_add(st, "d", "gw_id", gw->gw_id);
+	rpc->struct_add(st, "d", "gw_id", gw->flags);
+	rpc->struct_add(st, "d", "lcr_gw_id", gw->gw_id);
 	gw_name.s = gw->gw_name;
 	gw_name.len = gw->gw_name_len;
 	rpc->struct_add(st, "S", "gw_name", &gw_name);
@@ -558,7 +559,7 @@ static void lcr_print_stats(rpc_t *rpc, void *ctx)
 }
 
 #define CREATE_RPC_ENTRY(var1, var2)                                       \
-	snprintf(param, sizeof(param), "%s%u:%s = %lu", "gw", gw->gw_id, var1, \
+	snprintf(param, sizeof(param), "%s%u:%s = %lu", "gw", gw->flags, var1, \
 			var2);                                                         \
 	paramstr.s = param;                                                    \
 	paramstr.len = strlen(param);                                          \
