From: Victor Seva <vseva@sipwise.com>
Date: Fri, 30 Aug 2024 11:23:13 +0200
Subject: lost: add method parameter to http_connect() calls

---
 src/modules/lost/functions.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/modules/lost/functions.c b/src/modules/lost/functions.c
index d2c0f0c..565e134 100644
--- a/src/modules/lost/functions.c
+++ b/src/modules/lost/functions.c
@@ -188,7 +188,7 @@ int lost_held_function(struct sip_msg *_m, char *_con, char *_pidf, char *_url,
 	str geo = STR_NULL; /* return value geolocation uri */
 	str res = STR_NULL; /* return value pidf */
 	str err = STR_NULL; /* return value error */
-
+	str method = str_init("POST");
 	str url = STR_NULL;
 	str did = STR_NULL;
 	str que = STR_NULL;
@@ -292,7 +292,7 @@ int lost_held_function(struct sip_msg *_m, char *_con, char *_pidf, char *_url,
 		LM_DBG("using connection [%.*s]\n", con.len, con.s);
 
 		/* send via connection */
-		curl = httpapi.http_connect(_m, &con, NULL, &res, mtheld, &que);
+		curl = httpapi.http_connect(_m, &con, &method, NULL, &res, mtheld, &que);
 	} else {
 		/* we have no connection ... do a NAPTR lookup */
 		if(lost_parse_host(did.s, &host, &flag) > 0) {
@@ -819,6 +819,7 @@ int lost_function(struct sip_msg *_m, char *_con, char *_uri, char *_name,
 	str err = STR_NULL;	 /* return value error */
 
 	str tmp = STR_NULL;
+	str method = str_init("POST");
 	str url = STR_NULL;
 	str urn = STR_NULL;
 	str req = STR_NULL;
@@ -1093,7 +1094,7 @@ int lost_function(struct sip_msg *_m, char *_con, char *_uri, char *_name,
 		pkg_free(urlrep); /* clean up */
 		urlrep = NULL;
 	} else {
-		curl = httpapi.http_connect(_m, &con, NULL, &ret, mtlost, &req);
+		curl = httpapi.http_connect(_m, &con, &method, NULL, &ret, mtlost, &req);
 	}
 	/* only HTTP 2xx responses are accepted */
 	if(curl >= 300 || curl < 100) {
