From: Victor Seva <linuxmaniac@torreviejawireless.org>
Date: Tue, 31 Jan 2023 19:38:37 +0100
Subject: [PATCH] Makefile: support parallelism on install-modules-doc rule

---
 src/Makefile | 7 ++-----
 1 file changed, 2 insertions(+), 5 deletions(-)

diff --git a/src/Makefile b/src/Makefile
index f10a178..e343caa 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -473,13 +473,10 @@ install-$(1)-doc: modules.lst $(doc_prefix)/$(doc_dir)$(1)
 		if [ -n "$$$$r" -a -r $(1)/"$$$$r/Makefile" ]; then \
 			if [ -f $(1)/"$$$$r"/README ]; then \
 				$$(call try_err,\
-					$(INSTALL_TOUCH) $(doc_prefix)/$(doc_dir)$(1)/README ); \
+					$(INSTALL_TOUCH) $(doc_prefix)/$(doc_dir)$(1)/README."$$$$r" ); \
 				$$(call try_err,\
 					$(INSTALL_DOC)  $(1)/"$$$$r"/README  \
-									$(doc_prefix)/$(doc_dir)$(1)/README ); \
-				$$(call try_err,\
-					mv -f $(doc_prefix)/$(doc_dir)$(1)/README \
-							$(doc_prefix)/$(doc_dir)$(1)/README."$$$$r" ); \
+									$(doc_prefix)/$(doc_dir)$(1)/README."$$$$r" ); \
 			fi ; \
 		fi ; \
 	done; true
