--- a/modules/nathelper/nathelper.c
+++ b/modules/nathelper/nathelper.c
@@ -339,6 +339,7 @@
 static char *force_socket_str = 0;
 static pid_t mypid;
 static int sipping_flag = -1;
+static int natping_disable_flag = -1;
 static int natping_processes = 1;
 static int contact_only = 0;
 
@@ -413,6 +414,7 @@
 	{"sipping_from",          STR_PARAM, &sipping_from.s        },
 	{"sipping_method",        STR_PARAM, &sipping_method.s      },
 	{"sipping_bflag",         INT_PARAM, &sipping_flag          },
+	{"natping_disable_bflag", INT_PARAM, &natping_disable_flag  },
 	{"natping_processes",     INT_PARAM, &natping_processes     },
 	{"natping_socket",        STR_PARAM, &natping_socket        },
 	{"keepalive_timeout",     INT_PARAM, &nh_keepalive_timeout  },
@@ -678,6 +680,7 @@
 		}
 
 		sipping_flag = (sipping_flag==-1)?0:(1<<sipping_flag);
+		natping_disable_flag = (natping_disable_flag==-1)?0:(1<<natping_disable_flag);
 
 		/* set reply function if SIP natping is enabled */
 		if (sipping_flag) {
@@ -1998,6 +2001,9 @@
 		memcpy( &aorhash, cp, sizeof(aorhash));
 		cp = (char*)cp + sizeof(aorhash);
 
+		if ((flags & natping_disable_flag)) /* always 0 if natping_disable_flag not set */
+			continue;
+
 		/* determin the destination */
 		if ( path.len && (flags&sipping_flag)!=0 ) {
 			/* send to first URI in path */
