--- a/modules/presence_dfks/add_events.c
+++ b/modules/presence_dfks/add_events.c
@@ -16,9 +16,9 @@ static str pu_415_rpl  = str_init("Unsup
 static str unk_dev = str_init("<notKnown/>");
 static str content_type =str_init("application/x-as-feature-event+xml");
 // -4
-static str dnd_xml = str_init("<?xml version='1.0' encoding='ISO-8859-1'?><DoNotDisturbEvent><device>%s</device><doNotDisturbOn>%s</doNotDisturbOn></DoNotDisturbEvent>");
+static str dnd_xml = str_init("<?xml version=\"1.0\" encoding=\"ISO-8859-1\"?>\n<DoNotDisturbEvent xmlns=\"http://www.ecma-international.org/standards/ecma-323/csta/ed3\">\n<device>%s</device>\n<doNotDisturbOn>%s</doNotDisturbOn>\n</DoNotDisturbEvent>\n\r\n");
 // -8
-static str fwd_xml = str_init("<?xml version='1.0' encoding='ISO-8859-1'?><ForwardingEvent><device>%s</device><forwardingType>%s</forwardingType><forwardStatus>%s</forwardStatus><forwardTo>%s</forwardTo></ForwardingEvent>");
+static str fwd_xml = str_init("<?xml version='1.0' encoding='ISO-8859-1'?><ForwardingEvent xmlns=\"http://www.ecma-international.org/standards/ecma-323/csta/ed3\">\n<device>%s</device>\n<forwardingType>%s</forwardingType>\n<forwardStatus>%s</forwardStatus>\n<forwardTo>%s</forwardTo>\n</ForwardingEvent>\n\r\n");
 
 int dfks_add_events(void)
 {
