--- a/modules/pua_reginfo/usrloc_cb.c
+++ b/modules/pua_reginfo/usrloc_cb.c
@@ -223,7 +223,7 @@ void reginfo_usrloc_cb(ucontact_t* c, in
 	publ_info_t publ;
 	str content_type;
 	udomain_t * domain;
-	urecord_t * record;
+	urecord_t * record = NULL;
 	int res;
 	str uri = {NULL, 0};
 	str user = {NULL, 0};
@@ -332,6 +332,7 @@ error:
 		if(body->s) xmlFree(body->s);
 		pkg_free(body);
 	}
+	if(record) ul.release_urecord(record);
 
 	return;
 }	
--- a/modules/usrloc/udomain.c
+++ b/modules/usrloc/udomain.c
@@ -171,6 +171,7 @@ static inline void get_static_urecord(ud
 	r.aorhash = ul_get_aorhash(_aor);
 	r.domain = _d->name;
 	*_r = &r;
+	LM_DBG("urecord static[%p]\n", &r);
 }
 
 
@@ -1151,6 +1152,7 @@ int get_urecord(udomain_t* _d, str* _aor
 		r = db_load_urecord( ul_dbh, _d, _aor);
 		if (r) {
 			*_r = r;
+			LM_DBG("DB_ONLY urecord[%p]\n", r);
 			return 0;
 		}
 	}
@@ -1207,6 +1209,7 @@ int get_urecord_by_ruid(udomain_t* _d, u
 							&& !memcmp(c->ruid.s, _ruid->s, _ruid->len)) {
 						*_r = r;
 						*_c = c;
+						LM_DBG("DB_ONLY urecord[%p]\n", r);
 						return 0;
 					}
 					c = c->next;
--- a/modules/usrloc/urecord.c
+++ b/modules/usrloc/urecord.c
@@ -550,6 +550,7 @@ int db_delete_urecord_by_ruid(str *_tabl
 void release_urecord(urecord_t* _r)
 {
 	if (db_mode==DB_ONLY) {
+		LM_DBG("free urecord[%p]\n", _r);
 		free_urecord(_r);
 	} else if (_r->contacts == 0) {
 		mem_delete_urecord(_r->slot->d, _r);
--- a/modules/usrloc/ucontact.c
+++ b/modules/usrloc/ucontact.c
@@ -139,7 +139,7 @@ ucontact_t* new_ucontact(str* _dom, str*
 	ucontact_xavp_store(c);
 #endif
 
-
+	LM_DBG("new ucontact[%p]\n", c);
 	return c;
 error:
 	LM_ERR("no more shm memory\n");
@@ -176,6 +176,7 @@ void free_ucontact(ucontact_t* _c)
 #ifdef WITH_XAVP
 	if (_c->xavp) xavp_destroy_list(&_c->xavp);
 #endif
+	LM_DBG("free ucontact[%p]\n", _c);
 	shm_free( _c );
 }
 
--- a/modules/pua_reginfo/notify.c
+++ b/modules/pua_reginfo/notify.c
@@ -458,7 +458,7 @@ next_contact:
 			}
 		}
 next_registration:
-		// if (ul_record) ul.release_urecord(ul_record);		
+		if (ul_record) ul.release_urecord(ul_record);
 		/* Unlock the domain for this AOR: */
 		if (aor_key.len > 0)
 			ul.unlock_udomain(domain, &aor_key);
