From: Sipwise Development Team <support@sipwise.com>
Date: Thu, 10 Mar 2022 16:34:37 +0100
Subject: pv-headers-clone-branch-ignore-skip-header

---
 src/modules/pv_headers/pvh_xavp.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/src/modules/pv_headers/pvh_xavp.c b/src/modules/pv_headers/pvh_xavp.c
index b8e4f92..8b2acbf 100644
--- a/src/modules/pv_headers/pvh_xavp.c
+++ b/src/modules/pv_headers/pvh_xavp.c
@@ -515,8 +515,6 @@ int pvh_clone_branch_xavi(struct sip_msg *msg, str *xname)
 	}
 
 	do {
-		if(pvh_skip_header(&sub->name))
-			continue;
 		if(sub->val.type == SR_XTYPE_DATA)
 			continue;
 		if(pvh_xavi_append_value(&sub->name, &sub->val, &br_xavi->val.v.xavp)
