From: Sipwise Development Team <support@sipwise.com>
Date: Thu, 26 Mar 2020 10:06:46 +0100
Subject: registrar_add_path

===================================================================
---
 src/modules/registrar/lookup.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/modules/registrar/lookup.c b/src/modules/registrar/lookup.c
index 682eb7b..eeffaf6 100644
--- a/src/modules/registrar/lookup.c
+++ b/src/modules/registrar/lookup.c
@@ -160,6 +160,7 @@ int xavp_rcd_helper(ucontact_t* ptr)
 	str xname_received = { "received", 8};
 	str xname_contact = { "contact", 7};
 	str xname_expires = {"expires", 7};
+	str xname_path = {"path", 4};
 	sr_xval_t xval;
 
 	if(ptr==NULL) return -1;
@@ -190,6 +191,13 @@ int xavp_rcd_helper(ucontact_t* ptr)
 	xval.v.i = (int) (ptr->expires - time(0));
 	xavp_add_value(&xname_expires, &xval, xavp);
 
+	if(ptr->path.len > 0) {
+		memset(&xval, 0, sizeof(sr_xval_t));
+		xval.type = SR_XTYPE_STR;
+		xval.v.s = ptr->path;
+		xavp_add_value(&xname_path, &xval, xavp);
+	}
+
 	if(list==NULL) {
 		/* no reg_xavp_rcd xavp in root list - add it */
 		xval.type = SR_XTYPE_XAVP;
