--- a/src/modules/rtpengine/rtpengine.c
+++ b/src/modules/rtpengine/rtpengine.c
@@ -3354,9 +3354,6 @@
 	/* initialize some basic bencode items */
 	if(!extra_dict) {
 		ng_flags.dict = bencode_dictionary(bencbuf);
-		if(parse_by_module) {
-			ng_flags.flags = bencode_list(bencbuf);
-		}
 	} else {
 		ng_flags.dict = extra_dict;
 		ng_flags.flags = bencode_dictionary_get(ng_flags.dict, "flags");
@@ -3366,6 +3363,9 @@
 		}
 	}
 
+	if (!ng_flags.flags)
+		ng_flags.flags = bencode_list(bencbuf);
+
 	if(parse_by_module) {
 		ng_flags.received_from = bencode_list(bencbuf);
 	}
@@ -3440,9 +3440,6 @@
 		if(ng_flags.direction && ng_flags.direction->child)
 			bencode_dictionary_add(
 					ng_flags.dict, "direction", ng_flags.direction);
-		/* flags */
-		if(ng_flags.flags && ng_flags.flags->child)
-			bencode_dictionary_add(ng_flags.dict, "flags", ng_flags.flags);
 		/* replace */
 		if(ng_flags.replace && ng_flags.replace->child)
 			bencode_dictionary_add(ng_flags.dict, "replace", ng_flags.replace);
@@ -3522,6 +3519,10 @@
 				sip_type_strings[msg->first_line.type]);
 	}
 
+	/* flags */
+	if(ng_flags.flags && ng_flags.flags->child)
+		bencode_dictionary_add(ng_flags.dict, "flags", ng_flags.flags);
+
 	/* add rtpp flags, if parsed by daemon */
 	if(!parse_by_module && flags)
 		bencode_dictionary_add_str(ng_flags.dict, "rtpp-flags", flags);
