--- a/src/modules/rtpengine/rtpengine.c
+++ b/src/modules/rtpengine/rtpengine.c
@@ -2992,6 +2992,8 @@
 			if(ng_flags->to && ng_flags->to_tag.s && ng_flags->to_tag.len)
 				bencode_dictionary_add_str(
 						ng_flags->dict, "to-tag", &ng_flags->to_tag);
+			if(ng_flags->flags)
+				bencode_list_add_string(ng_flags->flags, "directional");
 		}
 	} else if(op == OP_SUBSCRIBE_REQUEST || op == OP_UNSUBSCRIBE) {
 		bencode_dictionary_add_str(
