From: Sipwise Development Team <support@sipwise.com>
Date: Tue, 7 Jan 2025 19:10:06 +0100
Subject: rtpengine_always_add_totag_to_ng_message_when_rtpp_flags

---
 src/modules/rtpengine/rtpengine.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/modules/rtpengine/rtpengine.c b/src/modules/rtpengine/rtpengine.c
index e26d533..97eff5b 100644
--- a/src/modules/rtpengine/rtpengine.c
+++ b/src/modules/rtpengine/rtpengine.c
@@ -3179,7 +3179,7 @@ static bencode_item_t *rtpp_function_call(bencode_buffer_t *bencbuf,
 	 */
 
 	/* affects to-tag parsing */
-	ng_flags.to = (op == OP_DELETE) ? 0 : 1;
+	ng_flags.to = (!parse_by_module || op != OP_DELETE) ? 1 : 0;
 
 	/* module specific parsing */
 	if(parse_by_module && flags && parse_flags(&ng_flags, msg, &op, flags->s))
