From: Victor Seva <linuxmaniac@torreviejawireless.org>
Date: Tue, 20 Sep 2016 18:28:20 +0200
Subject: sca: sca_call_info_update() fix param calls

---
 modules/sca/sca.c | 27 +++++++++++++++++++++++----
 1 file changed, 23 insertions(+), 4 deletions(-)

diff --git a/modules/sca/sca.c b/modules/sca/sca.c
index 92b174a..f46fb1b 100644
--- a/modules/sca/sca.c
+++ b/modules/sca/sca.c
@@ -52,19 +52,23 @@ static int		sca_mod_init( void );
 static int		sca_child_init( int );
 static void		sca_mod_destroy( void );
 static int		sca_set_config( sca_mod * );
+static int sca_call_info_update_0_f(sip_msg_t* msg);
+static int sca_call_info_update_1_f(sip_msg_t* msg, char*);
+static int sca_call_info_update_2_f(sip_msg_t* msg, char* , char*);
+static int sca_call_info_update_3_f(sip_msg_t* msg, char* , char*, char *);
 int fixup_ciu(void **, int);
 int fixup_free_ciu(void **param, int param_no);
 
 /* EXPORTED COMMANDS */
 static cmd_export_t	cmds[] = {
     { "sca_handle_subscribe", (cmd_function)sca_handle_subscribe, 0, NULL, NULL, REQUEST_ROUTE },
-    { "sca_call_info_update", (cmd_function)sca_call_info_update, 0, NULL, NULL,
+    { "sca_call_info_update", (cmd_function)sca_call_info_update_0_f, 0, NULL, NULL,
 	REQUEST_ROUTE | FAILURE_ROUTE | ONREPLY_ROUTE },
-    {"sca_call_info_update", (cmd_function)sca_call_info_update, 1,
+    {"sca_call_info_update", (cmd_function)sca_call_info_update_1_f, 1,
         fixup_ciu, fixup_free_ciu, REQUEST_ROUTE | FAILURE_ROUTE | ONREPLY_ROUTE},
-    {"sca_call_info_update", (cmd_function)sca_call_info_update, 2,
+    {"sca_call_info_update", (cmd_function)sca_call_info_update_2_f, 2,
         fixup_ciu, fixup_free_ciu, REQUEST_ROUTE | FAILURE_ROUTE | ONREPLY_ROUTE},
-    {"sca_call_info_update", (cmd_function)sca_call_info_update, 3,
+    {"sca_call_info_update", (cmd_function)sca_call_info_update_3_f, 3,
         fixup_ciu, fixup_free_ciu, REQUEST_ROUTE | FAILURE_ROUTE | ONREPLY_ROUTE},
     {0, 0, 0, 0, 0, 0},
 };
@@ -393,6 +397,21 @@ sca_mod_destroy( void )
     sca_db_disconnect();
 }
 
+static int sca_call_info_update_0_f(sip_msg_t* msg) {
+   return sca_call_info_update(msg, NULL, NULL, NULL);
+}
+static int sca_call_info_update_1_f(sip_msg_t* msg, char* p1) {
+   return sca_call_info_update(msg, p1, NULL, NULL);
+}
+static int sca_call_info_update_2_f(sip_msg_t* msg, char* p1, char* p2) {
+   return sca_call_info_update(msg, p1, p2, NULL);
+}
+static int sca_call_info_update_3_f(sip_msg_t* msg,
+   char* p1, char* p2, char * p3)
+{
+   return sca_call_info_update(msg, p1, p2, p3);
+}
+
 int fixup_ciu(void **param, int param_no)
 {
    switch (param_no) {
