From: Sipwise Development Team <support@sipwise.com>
Date: Tue, 4 Aug 2026 11:24:57 +0200
Subject: siputils_tel2sip_check_uri

---
 src/modules/siputils/checks.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/src/modules/siputils/checks.c b/src/modules/siputils/checks.c
index 2822e24..6a8d162 100644
--- a/src/modules/siputils/checks.c
+++ b/src/modules/siputils/checks.c
@@ -583,6 +583,7 @@ int tel2sip(struct sip_msg *_msg, char *_uri, char *_hostpart, char *_res)
 	int i, j, in_tel_parameters = 0;
 	pv_spec_t *res;
 	pv_value_t res_val;
+	struct sip_uri parsed_check;
 
 	/* get parameters */
 	if(get_str_fparam(&uri, _msg, (fparam_t *)_uri) < 0) {
@@ -647,6 +648,11 @@ int tel2sip(struct sip_msg *_msg, char *_uri, char *_hostpart, char *_res)
 	/* tel_uri is not needed anymore */
 	pkg_free(tel_uri.s);
 
+	if(parse_uri(sip_uri.s, sip_uri.len, &parsed_check) < 0) {
+		pkg_free(sip_uri.s);
+		return -1;
+	}
+
 	/* set result pv value and write sip uri to result pv */
 	res_val.rs = sip_uri;
 	res_val.flags = PV_VAL_STR;
@@ -761,6 +767,7 @@ int tel2sip2(struct sip_msg *_msg, char *_uri, char *_hostpart, char *_res)
 	pv_value_t res_val;
 	char *tmp_ptr = NULL;
 	tel_param_t params[MAX_TEL_PARAMS];
+	struct sip_uri parsed_check;
 
 	/* get parameters */
 	if(get_str_fparam(&uri, _msg, (fparam_t *)_uri) < 0) {
@@ -873,6 +880,11 @@ int tel2sip2(struct sip_msg *_msg, char *_uri, char *_hostpart, char *_res)
 	/* tel_uri is not needed anymore */
 	pkg_free(tel_uri.s);
 
+	if(parse_uri(sip_uri.s, sip_uri.len, &parsed_check) < 0) {
+		pkg_free(sip_uri.s);
+		return -1;
+	}
+
 	sip_uri.len = strlen(sip_uri.s);
 
 	/* set result pv value and write sip uri to result pv */
