--- /dev/null
+++ b/modules/tcap/Makefile
@@ -0,0 +1,9 @@
+include ../../Makefile.defs
+auto_gen=
+NAME=tcap.so
+
+DEFS+=-DKAMAILIO_MOD_INTERFACE -D_GNU_SOURCE
+DEFS+=-I/usr/include/tcap
+LIBS+=-ltcap
+
+include ../../Makefile.modules
--- /dev/null
+++ b/modules/tcap/tcap_mod.c
@@ -0,0 +1,37 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <tcap/libtcap.h>
+#include "../../sr_module.h"
+#include "../../mod_fix.h"
+#include "../../pvar.h"
+
+MODULE_VERSION
+
+static int tcap(sip_msg_t *msg, char *su, char *sq);
+
+static cmd_export_t cmds[] = {
+	{"tcap", (cmd_function)tcap, 2, fixup_spve_spve, 0,
+	REQUEST_ROUTE | FAILURE_ROUTE},
+	{0,}
+};
+
+struct module_exports exports = {
+	"tcap",
+	DEFAULT_DLFLAGS,
+	cmds,
+	0,
+	0,
+	0,
+	0,
+	0,
+	0,
+	0,
+	0,
+	0
+};
+
+static int tcap(sip_msg_t *msg, char *su, char *sq) {
+	return 0;
+}
--- a/Makefile.groups
+++ b/Makefile.groups
@@ -13,7 +13,7 @@
 				   mediaproxy mi_datagram mi_fifo mi_rpc mqueue \
 				   nat_traversal nathelper path pike pv ratelimit rr rtimer \
 				   rtpproxy sanity sdpops siputils sl statistics textops \
-				   textopsx tm tmx topoh xlog rtpproxy-ng stun sipt
+				   textopsx tm tmx topoh xlog rtpproxy-ng stun sipt tcap
 
 # - extra used modules, with no extra dependency
 mod_list_extra=avp auth_diameter call_control cnxcc dmq domainpolicy msrp pdb \
--- a/debian/control
+++ b/debian/control
@@ -37,6 +37,7 @@
  libxml2-dev,
  libxmlrpc-core-c3-dev,
  ngcp-libswrate-dev (>= 1.1.2+0~1364315461.svn13284.127),
+ ngcp-libtcap-dev,
  openssl,
  python,
  python-dev,
