From: Sipwise Development Team <support@sipwise.com>
Date: Wed, 5 Aug 2026 13:02:33 +0200
Subject: tlsa_dynamic_linking_libz

---
 src/modules/tlsa/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/modules/tlsa/Makefile b/src/modules/tlsa/Makefile
index adfe80b..b453480 100644
--- a/src/modules/tlsa/Makefile
+++ b/src/modules/tlsa/Makefile
@@ -53,7 +53,7 @@ else
 		DEFS += -I$(LIBSSL_STATIC_SRCPATH)/include
 	endif
 	LIBS += $(LIBSSL_STATIC_SRCPATH)/libssl.a $(LIBSSL_STATIC_SRCPATH)/libcrypto.a \
-			 -l:libz.a -ldl
+			 -lz -ldl
 endif # ifneq ($(LIBSSL_STATIC_SRCLIB),yes)
 
 else # ifneq ($(SSL_BUILDER),)
