Index: kamailio/src/modules/presence/presentity.c
===================================================================
--- kamailio.orig/src/modules/presence/presentity.c
+++ kamailio/src/modules/presence/presentity.c
@@ -379,6 +379,10 @@ int delete_presentity_if_dialog_id_exist
 	int i = 0;
 	presentity_t old_presentity;
 
+	if (presentity->event->evp->type != EVENT_DIALOG) {
+		return 0;
+	}
+
 	query_cols[n_query_cols] = &str_domain_col;
 	query_ops[n_query_cols] = OP_EQ;
 	query_vals[n_query_cols].type = DB1_STR;
