--- a/modules/rtpproxy/rtpproxy.c
+++ b/modules/rtpproxy/rtpproxy.c
@@ -283,7 +283,8 @@
 static int alter_rtcp(struct sip_msg *msg, str *body, str *oldport, str *newport);
 static char *gencookie();
 static int rtpp_test(struct rtpp_node*, int, int);
-static int unforce_rtp_proxy_f(struct sip_msg *, char *, char *);
+static int unforce_rtp_proxy0_f(struct sip_msg *, char *, char *);
+static int unforce_rtp_proxy1_f(struct sip_msg *, char *, char *);
 static int force_rtp_proxy(struct sip_msg *, char *, char *, int, int);
 static int start_recording_f(struct sip_msg *, char *, char *);
 static int rtpproxy_answer1_f(struct sip_msg *, char *, char *);
@@ -349,10 +350,10 @@
 	{"set_rtp_proxy_set",  (cmd_function)set_rtp_proxy_set_f,    1,
 		fixup_set_id, 0,
 		ANY_ROUTE},
-	{"unforce_rtp_proxy",  (cmd_function)unforce_rtp_proxy_f,    0,
+	{"unforce_rtp_proxy",  (cmd_function)unforce_rtp_proxy0_f,    0,
 		0, 0,
 		ANY_ROUTE},
-	{"rtpproxy_destroy",   (cmd_function)unforce_rtp_proxy_f,    0,
+	{"rtpproxy_destroy",   (cmd_function)unforce_rtp_proxy0_f,    0,
 		0, 0,
 		ANY_ROUTE},
 	{"start_recording",    (cmd_function)start_recording_f,      0,
@@ -362,19 +363,19 @@
 		0, 0,
 		ANY_ROUTE},
 	{"rtpproxy_offer",	(cmd_function)rtpproxy_offer1_f,     1,
-		0, 0,
+		fixup_spve_null, 0,
 		ANY_ROUTE},
 	{"rtpproxy_offer",	(cmd_function)rtpproxy_offer2_f,     2,
-		0, 0,
+		fixup_spve_null, 0,
 		ANY_ROUTE},
 	{"rtpproxy_answer",	(cmd_function)rtpproxy_answer1_f,    0,
 		0, 0,
 		ANY_ROUTE},
 	{"rtpproxy_answer",	(cmd_function)rtpproxy_answer1_f,    1,
-		0, 0,
+		fixup_spve_null, 0,
 		ANY_ROUTE},
 	{"rtpproxy_answer",	(cmd_function)rtpproxy_answer2_f,    2,
-		0, 0,
+		fixup_spve_spve, 0,
 		ANY_ROUTE},
 	{"rtpproxy_stream2uac",(cmd_function)rtpproxy_stream2uac2_f, 2,
 		fixup_var_str_int, 0,
@@ -1638,17 +1639,30 @@
 }
 
 static int
-unforce_rtp_proxy_f(struct sip_msg* msg, char* str1, char* str2)
+unforce_rtp_proxy0_f(struct sip_msg* msg, char* flags, char* str2)
+{
+	char arg[1] = {'\0'};
+	return unforce_rtp_proxy1_f(msg, arg, str2);
+}
+
+static int
+unforce_rtp_proxy1_f(struct sip_msg* msg, char* param1, char* param2)
 {
 	str callid, from_tag, to_tag, viabranch;
 	char *cp;
 	int via = 0;
 	int ret;
 	struct rtpp_node *node;
+	str str1;
 	struct iovec v[1 + 4 + 3 + 2] = {{NULL, 0}, {"D", 1}, {" ", 1}, {NULL, 0}, {NULL, 0}, {NULL, 0}, {" ", 1}, {NULL, 0}, {" ", 1}, {NULL, 0}};
 						    /* 1 */   /* 2 */   /* 3 */    /* 4 */    /* 5 */    /* 6 */   /* 7 */    /* 8 */   /* 9 */
 
-	for (cp = str1; cp && *cp; cp++) {
+	if (param1)
+		get_str_fparam(&str1, msg, (fparam_t *) param1);
+	else
+		str1.s = NULL;
+
+	for (cp = str1.s; cp && *cp; cp++) {
 		switch (*cp) {
 			case '1':
 				via = 1;
@@ -1740,7 +1754,7 @@
 		return -1;
 
 	if(method==METHOD_CANCEL || method==METHOD_BYE)
-		return unforce_rtp_proxy_f(msg, 0, 0);
+		return unforce_rtp_proxy0_f(msg, flags, 0);
 
 	if(ip==NULL)
 	{
@@ -1766,13 +1780,13 @@
 					&& tmb.t_gett()!=T_UNDEFINED)
 				tmb.t_gett()->uas.request->msg_flags |= FL_SDP_BODY;
 			if(route_type==FAILURE_ROUTE)
-				return unforce_rtp_proxy_f(msg, 0, 0);
+				return unforce_rtp_proxy0_f(msg, flags, 0);
 			return force_rtp_proxy(msg, flags, (cp!=NULL)?newip:ip, 1,
 					(ip!=NULL)?1:0);
 		}
 	} else if(msg->first_line.type == SIP_REPLY) {
 		if(msg->first_line.u.reply.statuscode>=300)
-			return unforce_rtp_proxy_f(msg, 0, 0);
+			return unforce_rtp_proxy0_f(msg, flags, 0);
 		if(nosdp==0) {
 			if(method==METHOD_UPDATE)
 				return force_rtp_proxy(msg, flags, (cp!=NULL)?newip:ip, 0,
@@ -1814,16 +1828,26 @@
 {
         char *cp;
         char newip[IP_ADDR_MAX_STR_SIZE];
+        str flags;
 
         cp = ip_addr2a(&msg->rcv.dst_ip);
         strcpy(newip, cp);
-	return force_rtp_proxy(msg, str1, newip, 1, 0);
+
+        if (str1)
+            get_str_fparam(&flags, msg, (fparam_t *) str1);
+        else
+            flags.s = NULL;
+        return force_rtp_proxy(msg, flags.s, newip, 1, 0);
 }
 
 static int
 rtpproxy_offer2_f(struct sip_msg *msg, char *param1, char *param2)
 {
-	return force_rtp_proxy(msg, param1, param2, 1, 1);
+        str flags, new_ip;
+
+        get_str_fparam(&flags, msg, (fparam_t *) param1);
+        get_str_fparam(&new_ip, msg, (fparam_t *) param2);
+        return force_rtp_proxy(msg, flags.s, new_ip.s, 1, 1);
 }
 
 static int
@@ -1831,6 +1855,7 @@
 {
         char *cp;
         char newip[IP_ADDR_MAX_STR_SIZE];
+        str flags;
 
 	if (msg->first_line.type == SIP_REQUEST)
 		if (msg->first_line.u.request.method_value != METHOD_ACK)
@@ -1838,18 +1863,26 @@
 
         cp = ip_addr2a(&msg->rcv.dst_ip);
         strcpy(newip, cp);
-	return force_rtp_proxy(msg, str1, newip, 0, 0);
+
+        if (str1)
+            get_str_fparam(&flags, msg, (fparam_t *) str1);
+        else
+            flags.s = NULL;
+        return force_rtp_proxy(msg, flags.s, newip, 0, 0);
 }
 
 static int
 rtpproxy_answer2_f(struct sip_msg *msg, char *param1, char *param2)
 {
+	str flags, new_ip;
 
 	if (msg->first_line.type == SIP_REQUEST)
 		if (msg->first_line.u.request.method_value != METHOD_ACK)
 			return -1;
 
-	return force_rtp_proxy(msg, param1, param2, 0, 1);
+	get_str_fparam(&flags, msg, (fparam_t *) param1);
+	get_str_fparam(&new_ip, msg, (fparam_t *) param2);
+	return force_rtp_proxy(msg, flags.s, new_ip.s, 0, 1);
 }
 
 
--- a/modules/rtpproxy/README
+++ b/modules/rtpproxy/README
@@ -70,7 +70,7 @@
 
         6. Exported Pseudo Variables
 
-              6.1. $rtpstat
+              6.1. $rtpstart
 
         7. MI Commands
 
@@ -95,7 +95,7 @@
    1.12. rtpproxy_manage usage
    1.13. rtpproxy_stream2xxx usage
    1.14. start_recording usage
-   1.15. $rtpstat-Usage
+   1.15. $rtpstart-Usage
    1.16. nh_enable_rtpp usage
    1.17. nh_show_rtpp usage
 
@@ -136,7 +136,7 @@
 
    6. Exported Pseudo Variables
 
-        6.1. $rtpstat
+        6.1. $rtpstart
 
    7. MI Commands
 
@@ -588,18 +588,18 @@
 
 6. Exported Pseudo Variables
 
-   6.1. $rtpstat
+   6.1. $rtpstart
 
-6.1. $rtpstat
+6.1. $rtpstart
 
    Returns the RTP-Statistics from the RTP-Proxy. The RTP-Statistics from
    the RTP-Proxy are provided as a string and it does contain several
    packet-counters. The statistics must be retrieved before the session is
    deleted (before unforce_rtpproxy).
 
-   Example 1.15. $rtpstat-Usage
+   Example 1.15. $rtpstart-Usage
 ...
-    append_hf("X-RTP-Statistics: $rtpstat\r\n");
+    append_hf("X-RTP-Statistics: $rtpstart\r\n");
 ...
 
 7. MI Commands
