Index: modules_k/usrloc/ul_mi.c
===================================================================
--- modules_k/usrloc/ul_mi.c	(revisión: 9238)
+++ modules_k/usrloc/ul_mi.c	(revisión: 9239)
@@ -525,8 +525,10 @@
 	if (str2q( &ci.q, node->value.s, node->value.len) < 0)
 		goto bad_syntax;
 
-	/* unused value (param 6) FIXME */
+	/* path value (param 6) */
 	node = node->next;
+	if(strncmp(node->value.s, "0", 1) != 0 && node->value.len > 1)
+		ci.path = &node->value;
 
 	/* flags value (param 7) */
 	node = node->next;
Index: utils/kamctl/kamctl
===================================================================
--- utils/kamctl/kamctl	(revisión: 9238)
+++ utils/kamctl/kamctl	(revisión: 9239)
@@ -2201,10 +2201,17 @@
 				UL_EXPIRES=0
 				UL_FLAGS=0
 				BR_FLAGS=0
+				UL_PATH=0
 			elif [ $# -eq 4 ] ; then
 				UL_EXPIRES=$4
 				UL_FLAGS=0
 				BR_FLAGS=0
+				UL_PATH=0
+			elif [ $# -eq 5 ] ; then
+				UL_EXPIRES=$4
+				UL_FLAGS=0
+				BR_FLAGS=0
+				UL_PATH="$5"
 			else
 				usage_usrloc
 				exit 1
@@ -2237,7 +2244,7 @@
 			fi
 
 			$CTLCMD ul_add "$USRLOC_TABLE" "$OSERUSER@$OSERDOMAIN" "$2" \
-"$UL_EXPIRES" "1.00" "0" "$UL_FLAGS" "$BR_FLAGS" "$ALL_METHODS"
+"$UL_EXPIRES" "1.00" "$UL_PATH" "$UL_FLAGS" "$BR_FLAGS" "$ALL_METHODS"
 			exit $?
 			;;
 		rm)
