From d5c4dda7932e8efe018beb77f8ca2e29a44aa185 Mon Sep 17 00:00:00 2001
From: Daniel-Constantin Mierla <miconda@gmail.com>
Date: Thu, 30 May 2013 17:59:01 +0200
Subject: [PATCH] core: don't free pvar spec - is reference to cache

- reported by Peter Dunkley
---
 cfg.y |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cfg.y b/cfg.y
index 0e5f8b1..5f88d13 100644
--- a/cfg.y
+++ b/cfg.y
@@ -2696,7 +2696,7 @@ rval: intno			{$$=mk_rve_rval(RV_INT, (void*)$1); }
 	| STRING			{	s_tmp.s=$1; s_tmp.len=strlen($1);
 							$$=mk_rve_rval(RV_STR, &s_tmp); }
 	| attr_id_any		{$$=mk_rve_rval(RV_AVP, $1); pkg_free($1); }
-	| pvar				{$$=mk_rve_rval(RV_PVAR, $1); pkg_free($1); }
+	| pvar				{$$=mk_rve_rval(RV_PVAR, $1); }
 	| avp_pvar			{
 							switch($1->type){
 								case LV_AVP:
-- 
1.7.10.4

