From 1207edd603937e3e43ce20207bfeae0284f16489 Mon Sep 17 00:00:00 2001
From: Daniel-Constantin Mierla <miconda@gmail.com>
Date: Mon, 17 Jun 2013 14:48:52 +0200
Subject: [PATCH 2/2] pv: $xavp(name) is marked as PVT_XAVP type

---
 modules/pv/pv.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/modules/pv/pv.c
+++ b/modules/pv/pv.c
@@ -82,7 +82,7 @@
 		pv_parse_snd_name, 0, 0, 0 },
 #ifdef WITH_XAVP
 	{ {"xavp", sizeof("xavp")-1}, /* xavp */
-		PVT_OTHER, pv_get_xavp, pv_set_xavp,
+		PVT_XAVP, pv_get_xavp, pv_set_xavp,
 		pv_parse_xavp_name, 0, 0, 0 },
 #endif
 
