From 802771362eee707ecb55c2a907f96d2dbaff620e Mon Sep 17 00:00:00 2001
From: Victor Seva <linuxmaniac@torreviejawireless.org>
Date: Mon, 28 Apr 2014 16:20:57 +0200
Subject: [PATCH] core: add str_append helper function

---
 str.c | 45 +++++++++++++++++++++++++++++++++++++++++++++
 str.h |  9 +++++++++
 2 files changed, 54 insertions(+)
 create mode 100644 str.c

diff --git a/str.c b/str.c
new file mode 100644
index 0000000..913dc7b
--- /dev/null
+++ b/str.c
@@ -0,0 +1,45 @@
+/**
+ * $Id$
+ *
+ * Copyright (C) 2014 Victor Seva <vseva@sipwise.com>
+ *
+ * This file is part of kamailio, a free SIP server.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <string.h>
+#include "str.h"
+#include "mem/mem.h"
+
+int str_append(str *orig, str *suffix, str *dest)
+{
+	if(orig == NULL || suffix == NULL || suffix->len == 0 || dest == NULL)
+	{
+		LM_ERR("wrong parameters\n");
+		return -1;
+	}
+	dest->len = orig->len + suffix->len;
+	dest->s = pkg_malloc(sizeof(char)*dest->len);
+	if(dest->s==NULL)
+	{
+		LOG(L_ERR, "memory allocation failure\n");
+		return -1;
+	}
+	if(orig->len>0)
+	{
+		memcpy(dest->s, orig->s, orig->len);
+	}
+	memcpy(dest->s+orig->len, suffix->s, suffix->len);
+	return 0;
+}
diff --git a/str.h b/str.h
index 1985730..5414be0 100644
--- a/str.h
+++ b/str.h
@@ -121,4 +121,13 @@ typedef struct _str str;
 
 /** @} */
 
+/** Appends a sufffix
+ * @param orig is the original string
+ * @param suffix is the suffix string
+ * @param dest is the result ::str of appending suffix to orig
+ * @return 0 if ok -1 if error
+ * remember to free the dest->s private memory
+ */
+int str_append(str *orig, str *suffix, str *dest);
+
 #endif
-- 
2.0.0.rc0

