From c5d76e3282fa49e53ee1bc6e25c6a0315f0aadcb Mon Sep 17 00:00:00 2001
From: lazedo <luis.azedo@factorlusitano.com>
Date: Tue, 24 Jan 2017 01:11:35 +0000
Subject: [PATCH] tm: apply new updates with Via with updated request

---
 modules/tm/uac.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/modules/tm/uac.c b/modules/tm/uac.c
index 8c3a485a8..93ed0c912 100644
--- a/modules/tm/uac.c
+++ b/modules/tm/uac.c
@@ -426,7 +426,7 @@ static inline int t_uac_prepare(uac_req_t *uac_r,
 
 					LM_DBG("apply new updates with Via to sip msg\n");
 					buf1 = build_req_buf_from_sip_req(&lreq,
-							(unsigned int*)&buf_len1, &dst, BUILD_IN_SHM);
+							(unsigned int*)&buf_len1, &request->dst, BUILD_IN_SHM);
 					if (likely(buf1)){
 						shm_free(buf);
 						buf = buf1;
-- 
2.11.0

