From 7b011b32be9bca6cf618e2eecd8364dce4041af6 Mon Sep 17 00:00:00 2001
From: Luis Azedo <luis@2600hz.com>
Date: Fri, 27 Mar 2015 01:36:45 +0000
Subject: [PATCH] registrar: fix wrong var name

---
 modules/registrar/lookup.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/modules/registrar/lookup.c
+++ b/modules/registrar/lookup.c
@@ -588,7 +588,7 @@
 	}
 
 	if (res == 0) {
-		LM_DBG("searching with match flags (%d,%d)\n", match_flag, reg_match_flag_param);
+		LM_DBG("searching with match flags (%d,%d)\n", match_flag, match_action_flag);
 		if(reg_xavp_cfg.s!=NULL) {
 
 			if((match_flag & 1)
