From fed41e7bd30a1c4cd925b7f0c73d8aaa44722955 Mon Sep 17 00:00:00 2001
From: Victor Seva <linuxmaniac@torreviejawireless.org>
Date: Tue, 25 Aug 2015 15:43:29 +0200
Subject: [PATCH] core: fix routename for event_route

---
 cfg.lex | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/cfg.lex b/cfg.lex
index 95edfdb..147f597 100644
--- a/cfg.lex
+++ b/cfg.lex
@@ -605,23 +605,23 @@ IMPORTFILE      "import_file"
 <INITIAL>{ISAVPFLAGSET}	{ count(); yylval.strval=yytext; return ISAVPFLAGSET; }
 <INITIAL>{AVPFLAGS_DECL}	{ count(); yylval.strval=yytext; return AVPFLAGS_DECL; }
 <INITIAL>{MSGLEN}	{ count(); yylval.strval=yytext; return MSGLEN; }
-<INITIAL>{ROUTE}	{ count(); default_routename="DEFAULT_ROUTE";
+<INITIAL>{ROUTE}	{ count(); routename=NULL; default_routename="DEFAULT_ROUTE";
 						yylval.strval=yytext; return ROUTE; }
-<INITIAL>{ROUTE_REQUEST}	{ count(); default_routename="DEFAULT_ROUTE";
+<INITIAL>{ROUTE_REQUEST}	{ count(); routename=NULL; default_routename="DEFAULT_ROUTE";
 								yylval.strval=yytext; return ROUTE_REQUEST; }
-<INITIAL>{ROUTE_ONREPLY}	{ count(); default_routename="DEFAULT_ONREPLY";
+<INITIAL>{ROUTE_ONREPLY}	{ count(); routename=NULL; default_routename="DEFAULT_ONREPLY";
 								yylval.strval=yytext;
 								return ROUTE_ONREPLY; }
-<INITIAL>{ROUTE_REPLY}	{ count(); default_routename="DEFAULT_ONREPLY";
+<INITIAL>{ROUTE_REPLY}	{ count(); routename=NULL; default_routename="DEFAULT_ONREPLY";
 							yylval.strval=yytext; return ROUTE_REPLY; }
-<INITIAL>{ROUTE_FAILURE}	{ count(); default_routename="DEFAULT_FAILURE";
+<INITIAL>{ROUTE_FAILURE}	{ count(); routename=NULL; default_routename="DEFAULT_FAILURE";
 								yylval.strval=yytext;
 								return ROUTE_FAILURE; }
-<INITIAL>{ROUTE_BRANCH} { count(); default_routename="DEFAULT_BRANCH";
+<INITIAL>{ROUTE_BRANCH} { count(); routename=NULL; default_routename="DEFAULT_BRANCH";
 							yylval.strval=yytext; return ROUTE_BRANCH; }
-<INITIAL>{ROUTE_SEND} { count(); default_routename="DEFAULT_SEND";
+<INITIAL>{ROUTE_SEND} { count(); routename=NULL; default_routename="DEFAULT_SEND";
 							yylval.strval=yytext; return ROUTE_SEND; }
-<INITIAL>{ROUTE_EVENT} { count(); default_routename="DEFAULT_EVENT";
+<INITIAL>{ROUTE_EVENT} { count(); routename=NULL; default_routename="DEFAULT_EVENT";
 							yylval.strval=yytext;
 							state=EVRT_NAME_S; BEGIN(EVRTNAME);
 							return ROUTE_EVENT; }
@@ -630,6 +630,7 @@ IMPORTFILE      "import_file"
 <EVRTNAME>{EVENT_RT_NAME}	{ count();
 								addstr(&s_buf, yytext, yyleng);
 								yylval.strval=s_buf.s;
+								routename=s_buf.s;
 								memset(&s_buf, 0, sizeof(s_buf));
 								return EVENT_RT_NAME; }
 <EVRTNAME>{RBRACK}          { count();
-- 
2.5.0

