From c05d840d541b87a45962d8f9b39d85f608ee533f Mon Sep 17 00:00:00 2001
From: Victor Seva <linuxmaniac@torreviejawireless.org>
Date: Tue, 25 Aug 2015 15:43:29 +0200
Subject: [PATCH] core: fix routename for event_route

---
 cfg.lex | 16 ++++++++--------
 cfg.y   |  1 +
 2 files changed, 9 insertions(+), 8 deletions(-)

diff --git a/cfg.lex b/cfg.lex
index 95edfdb..6cad508 100644
--- a/cfg.lex
+++ b/cfg.lex
@@ -605,23 +605,23 @@ IMPORTFILE      "import_file"
 <INITIAL>{ISAVPFLAGSET}	{ count(); yylval.strval=yytext; return ISAVPFLAGSET; }
 <INITIAL>{AVPFLAGS_DECL}	{ count(); yylval.strval=yytext; return AVPFLAGS_DECL; }
 <INITIAL>{MSGLEN}	{ count(); yylval.strval=yytext; return MSGLEN; }
-<INITIAL>{ROUTE}	{ count(); default_routename="DEFAULT_ROUTE";
+<INITIAL>{ROUTE}	{ count(); routename=NULL; default_routename="DEFAULT_ROUTE";
 						yylval.strval=yytext; return ROUTE; }
-<INITIAL>{ROUTE_REQUEST}	{ count(); default_routename="DEFAULT_ROUTE";
+<INITIAL>{ROUTE_REQUEST}	{ count(); routename=NULL; default_routename="DEFAULT_ROUTE";
 								yylval.strval=yytext; return ROUTE_REQUEST; }
-<INITIAL>{ROUTE_ONREPLY}	{ count(); default_routename="DEFAULT_ONREPLY";
+<INITIAL>{ROUTE_ONREPLY}	{ count(); routename=NULL; default_routename="DEFAULT_ONREPLY";
 								yylval.strval=yytext;
 								return ROUTE_ONREPLY; }
-<INITIAL>{ROUTE_REPLY}	{ count(); default_routename="DEFAULT_ONREPLY";
+<INITIAL>{ROUTE_REPLY}	{ count(); routename=NULL; default_routename="DEFAULT_ONREPLY";
 							yylval.strval=yytext; return ROUTE_REPLY; }
-<INITIAL>{ROUTE_FAILURE}	{ count(); default_routename="DEFAULT_FAILURE";
+<INITIAL>{ROUTE_FAILURE}	{ count(); routename=NULL; default_routename="DEFAULT_FAILURE";
 								yylval.strval=yytext;
 								return ROUTE_FAILURE; }
-<INITIAL>{ROUTE_BRANCH} { count(); default_routename="DEFAULT_BRANCH";
+<INITIAL>{ROUTE_BRANCH} { count(); routename=NULL; default_routename="DEFAULT_BRANCH";
 							yylval.strval=yytext; return ROUTE_BRANCH; }
-<INITIAL>{ROUTE_SEND} { count(); default_routename="DEFAULT_SEND";
+<INITIAL>{ROUTE_SEND} { count(); routename=NULL; default_routename="DEFAULT_SEND";
 							yylval.strval=yytext; return ROUTE_SEND; }
-<INITIAL>{ROUTE_EVENT} { count(); default_routename="DEFAULT_EVENT";
+<INITIAL>{ROUTE_EVENT} { count(); routename=NULL; default_routename="DEFAULT_EVENT";
 							yylval.strval=yytext;
 							state=EVRT_NAME_S; BEGIN(EVRTNAME);
 							return ROUTE_EVENT; }
diff --git a/cfg.y b/cfg.y
index 154cb44..d12e25f 100644
--- a/cfg.y
+++ b/cfg.y
@@ -1966,6 +1966,7 @@ event_route_stm: ROUTE_EVENT LBRACK EVENT_RT_NAME RBRACK LBRACE actions RBRACE {
 			yyerror("duplicate route");
 			YYABORT;
 		}
+		routename=$3;
 		push($6, &event_rt.rlist[i_tmp]);
 	}
 
-- 
2.5.0

