From 1bb1c30985d2eca6fabe4b60bb16edce8f81f385 Mon Sep 17 00:00:00 2001
From: Richard Fuchs <rfuchs@sipwise.com>
Date: Tue, 9 Apr 2013 16:16:10 -0400
Subject: [PATCH] fix string length matching

---
 modules_k/pv/pv_trans.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/modules_k/pv/pv_trans.c
+++ b/modules_k/pv/pv_trans.c
@@ -1749,10 +1749,10 @@
 	} else if(name.len==11 && strncasecmp(name.s, "decode.hexa", 11)==0) {
 		t->subtype = TR_S_DECODEHEXA;
 		goto done;
-	} else if(name.len==11 && strncasecmp(name.s, "encode.base64", 13)==0) {
+	} else if(name.len==13 && strncasecmp(name.s, "encode.base64", 13)==0) {
 		t->subtype = TR_S_ENCODEBASE64;
 		goto done;
-	} else if(name.len==11 && strncasecmp(name.s, "decode.base64", 13)==0) {
+	} else if(name.len==13 && strncasecmp(name.s, "decode.base64", 13)==0) {
 		t->subtype = TR_S_DECODEBASE64;
 		goto done;
 	} else if(name.len==13 && strncasecmp(name.s, "escape.common", 13)==0) {
