--- a/modules/registrar/lookup.c
+++ b/modules/registrar/lookup.c
@@ -706,7 +706,7 @@ int registered4(struct sip_msg* _m, udom
 
 			if(ptr->xavp!=NULL && match_action_flag == 1) {
 				sr_xavp_t *xavp = xavp_clone_level_nodata(ptr->xavp);
-				if(xavp_add(xavp, NULL)<0) {
+				if(xavp_insert(xavp, 0, NULL)<0) {
 					LM_ERR("error adding xavp for %.*s after successful match\n", aor.len, ZSW(aor.s));
 					xavp_destroy_list(&xavp);
 				}
