--- a/modules/acc/acc_logic.c
+++ b/modules/acc/acc_logic.c
@@ -361,27 +361,40 @@ static inline int should_acc_reply(struc
 {
     unsigned int i;
 
+    DBG("DEBUG: probing acc state - code: %d flags: 0x%x\n", code,
+			(req)?req->flags:0);
 	/* negative transactions reported otherwise only if explicitly 
 	 * demanded */
 
     if (code >= 300) {
-	if (!is_failed_acc_on(req)) return 0;
-	i = 0;
-	while (failed_filter[i] != 0) {
-	    if (failed_filter[i] == code) return 0;
-	    i++;
-	}
-	return 1;
+		if (!is_failed_acc_on(req)) {
+			DBG("DEBUG: failed acc is off\n");
+			return 0;
+		}
+		i = 0;
+		while (failed_filter[i] != 0) {
+		    if (failed_filter[i] == code) {
+		    	DBG("DEBUG: acc code in filter: %d\n", code);
+		    	return 0;
+		    }
+		    i++;
+		}
+		return 1;
     }
 
-    if ( !is_acc_on(req) )
-	return 0;
+    if ( !is_acc_on(req) ) {
+    	DBG("DEBUG: acc is off\n");
+    	return 0;
+    }
 	
     if ( code<200 && !(early_media &&
 		       parse_headers(rpl,HDR_CONTENTLENGTH_F, 0) == 0 &&
-		       rpl->content_length && get_content_length(rpl) > 0))
-	return 0;
+    		   rpl->content_length && get_content_length(rpl) > 0)) {
+		DBG("DEBUG: early media acc is off\n");
+		return 0;
+	}
 
+	DBG("DEBUG: acc is on\n");
     return 1; /* seed is through, we will account this reply */
 }
 
@@ -409,6 +422,7 @@ static inline void on_missed(struct cell
 	int flags_to_reset = 0;
 	int br = -1;
 
+	DBG("DEBUG: preparing to report the record\n");
 	/* get winning branch index, if set */
 	if (t->relayed_reply_branch>=0) {
 		br = t->relayed_reply_branch;
