From 08248bae17024246442046376a4350007a100025 Mon Sep 17 00:00:00 2001
From: Daniel-Constantin Mierla <miconda@gmail.com>
Date: Mon, 10 Nov 2014 15:42:39 +0100
Subject: [PATCH] tm: handle BRANCH_FAILURE_ROUTE in t_check_trans() test

---
 modules/tm/tm.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

--- a/modules/tm/tm.c
+++ b/modules/tm/tm.c
@@ -2110,9 +2110,10 @@
 	int ret;
 	
 	/* already processing a T */
-	if(get_route_type()==FAILURE_ROUTE
-			|| get_route_type()==BRANCH_ROUTE
-			|| get_route_type()==TM_ONREPLY_ROUTE) {
+	if(is_route_type(FAILURE_ROUTE)
+			|| is_route_type(BRANCH_ROUTE)
+			|| is_route_type(BRANCH_FAILURE_ROUTE)
+			|| is_route_type(TM_ONREPLY_ROUTE)) {
 		return 1;
 	}
 
