From: Victor Seva <vseva@sipwise.com>
Date: Wed, 27 May 2020 09:00:19 +0200
Subject: core: receive - reset xavi list after message processing

---
 src/core/receive.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/core/receive.c b/src/core/receive.c
index 550ec59..e23de0f 100644
--- a/src/core/receive.c
+++ b/src/core/receive.c
@@ -560,4 +560,5 @@ void ksr_msg_env_reset(void)
 {
 	reset_avps();
 	xavp_reset_list();
+	xavi_reset_list();
 }
