From 83a437144bad5d31f5ce1bfc10b03f613524852f Mon Sep 17 00:00:00 2001
From: lazedo <luis.azedo@factorlusitano.com>
Date: Fri, 18 Jan 2019 13:18:01 +0000
Subject: [PATCH] core: allow pp_define_get to be used by modules

---
 src/core/cfg.lex | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/core/cfg.lex b/src/core/cfg.lex
index 6e4d6971e..9912d111b 100644
--- a/src/core/cfg.lex
+++ b/src/core/cfg.lex
@@ -113,7 +113,7 @@
 		struct sr_yy_fname *next;
 	} *sr_yy_fname_list = 0;
 
-	static str  *pp_define_get(int len, const char * text);
+	str  *pp_define_get(int len, const char * text);
 	static int  pp_ifdef_type(int pos);
 	static void pp_ifdef_var(int len, const char * text);
 	static void pp_ifdef();
@@ -1856,7 +1856,7 @@ int pp_define_set(int len, char *text)
 	return 0;
 }
 
-static str *pp_define_get(int len, const char * text)
+str *pp_define_get(int len, const char * text)
 {
 	str var = {(char *)text, len};
 	int i;
-- 
2.20.1

