From b5f27fe25247e3605aeeb600de5a6b31081d6ac6 Mon Sep 17 00:00:00 2001
From: Daniel-Constantin Mierla <miconda@gmail.com>
Date: Thu, 24 Jan 2019 15:53:55 +0100
Subject: [PATCH] core: export pp_define_get() via ppcfg.h

---
 src/core/ppcfg.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/core/ppcfg.h b/src/core/ppcfg.h
index 432d76bcf..848505991 100644
--- a/src/core/ppcfg.h
+++ b/src/core/ppcfg.h
@@ -41,6 +41,7 @@ int pp_subst_run(char **data);
 int  pp_define(int len, const char *text);
 int  pp_define_set(int len, char *text);
 int  pp_define_set_type(int type);
+str *pp_define_get(int len, const char * text);
 
 void pp_ifdef_level_update(int val);
 void pp_ifdef_level_check(void);
-- 
2.20.1

