--- a/src/modules/db_redis/redis_dbase.c
+++ b/src/modules/db_redis/redis_dbase.c
@@ -1393,7 +1393,6 @@
         //db_redis_key_free(&type_keys);
         LM_DBG("+++ done with loop '%.*s'\n", k->key.len, k->key.s);
     }
-    pkg_free(query_v);
     db_redis_key_free(&type_keys);
     db_redis_key_free(&all_type_keys);
 
@@ -1598,7 +1597,8 @@
                 LM_ERR("Failed to add key to update query\n");
                 goto error;
             }
-            pkg_free(v.s);
+            if (v.s)
+                pkg_free(v.s);
         }
         update_queries++;
         if (db_redis_append_command_argv(con, query_v, 1) != REDIS_OK) {
@@ -1850,7 +1850,8 @@
             LM_ERR("Failed to add column value to insert query\n");
             goto error;
         }
-        pkg_free(v.s);
+        if (v.s)
+            pkg_free(v.s);
     }
 
     reply = db_redis_command_argv(con, query_v);
@@ -2148,4 +2149,4 @@
         return -1;
     db_free_result(_r);
     return 0;
-}
\ No newline at end of file
+}
