From c32f75cc1a02537474614f52cecfd5a19c56184d Mon Sep 17 00:00:00 2001
From: Daniel-Constantin Mierla <miconda@gmail.com>
Date: Fri, 9 Jan 2015 17:58:12 +0100
Subject: [PATCH] Makefile.groups: added grou for jansson modules

---
 Makefile.groups | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

--- a/Makefile.groups
+++ b/Makefile.groups
@@ -135,9 +135,12 @@
 # - modules depending on oracle library
 mod_list_oracle=db_oracle
 
-# - modules depending on json library
+# - modules depending on json (+libevent) library
 mod_list_json=json jsonrpc-c
 
+# - modules depending on jansson (+libevent) library
+mod_list_jansson=jansson janssonrpc-c
+
 # - modules depending on redis library
 mod_list_redis=ndb_redis
 
@@ -186,7 +189,7 @@
 			   $(mod_list_outbound) $(mod_list_osp) \
 			   $(mod_list_java) $(mod_list_dnssec) \
 			   $(mod_list_sctp) $(mod_list_autheph) \
-			   $(mod_list_gzcompress)) \
+			   $(mod_list_gzcompress)) $(mod_list_jansson)) \
 			   $(mod_list_lcrrate)
 
 
@@ -326,6 +329,9 @@
 # K json modules
 module_group_kjson=$(mod_list_json)
 
+# K jansson modules
+module_group_kjansson=$(mod_list_jansson)
+
 # pkg redis module
 module_group_kredis=$(mod_list_redis)
 
