From: Victor Seva <linuxmaniac@torreviejawireless.org>
Date: Thu, 4 Nov 2021 14:34:47 +0100
Subject: permissions: don't remove old data at the end of the reload process

---
 src/modules/permissions/trusted.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/src/modules/permissions/trusted.c b/src/modules/permissions/trusted.c
index 2f9c7d4..7714a32 100644
--- a/src/modules/permissions/trusted.c
+++ b/src/modules/permissions/trusted.c
@@ -59,7 +59,6 @@ int reload_trusted_table(void)
 	db_val_t* val;
 
 	struct trusted_list **new_hash_table;
-	struct trusted_list **old_hash_table;
 	int i;
 	int priority;
 
@@ -160,9 +159,7 @@ int reload_trusted_table(void)
 
 	perm_dbf.free_result(perm_db_handle, res);
 
-	old_hash_table = *perm_trust_table;
 	*perm_trust_table = new_hash_table;
-	empty_hash_table(old_hash_table);
 
 	LM_DBG("trusted table reloaded successfully.\n");
 
