From df472fa9bdaa1a796dd220c99a64b1539c8854b5 Mon Sep 17 00:00:00 2001
From: Victor Seva <linuxmaniac@torreviejawireless.org>
Date: Fri, 20 Sep 2013 11:07:14 +0200
Subject: [PATCH] pv: Move pv_xavp_name_t declaration to core pvar.h

---
 modules/pv/pv_xavp.h |    6 ------
 pvar.h               |   10 ++++++++++
 2 files changed, 10 insertions(+), 6 deletions(-)

--- a/modules/pv/pv_xavp.h
+++ b/modules/pv/pv_xavp.h
@@ -23,12 +23,6 @@
 
 #include "../../pvar.h"
 
-typedef struct _pv_xavp_name {
-	str name;
-	pv_spec_t index;
-	struct _pv_xavp_name *next;
-} pv_xavp_name_t;
-
 int pv_get_xavp(struct sip_msg *msg, pv_param_t *param,
 		pv_value_t *res);
 int pv_set_xavp(struct sip_msg* msg, pv_param_t *param,
--- a/pvar.h
+++ b/pvar.h
@@ -287,5 +287,15 @@
 
 int register_trans_mod(char *mod_name, tr_export_t *items);
 
+
+/**
+ * XAVP
+ */
+typedef struct _pv_xavp_name {
+	str name;
+	pv_spec_t index;
+	struct _pv_xavp_name *next;
+} pv_xavp_name_t;
+
 #endif
 
