From: Surendra Tiwari <surendratiwari3@gmail.com>
Date: Wed, 9 Jan 2019 08:57:12 +0100
Subject: [PATCH] rtpengine: exported rtpengine_query() variants to KEMI
 framework

- GH #1792
---
 src/modules/rtpengine/rtpengine.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/src/modules/rtpengine/rtpengine.c b/src/modules/rtpengine/rtpengine.c
index 20df725..77a0af9 100644
--- a/src/modules/rtpengine/rtpengine.c
+++ b/src/modules/rtpengine/rtpengine.c
@@ -3986,6 +3986,16 @@ static int ki_rtpengine_query(sip_msg_t *msg, str *flags)
         return rtpengine_rtpp_set_wrap(msg, rtpengine_query_wrap, flags->s, 1, OP_ANY);
 }
 
+static int ki_rtpengine_query0(sip_msg_t *msg)
+{       
+        return rtpengine_rtpp_set_wrap(msg, rtpengine_query_wrap, NULL, 1);
+}
+
+static int ki_rtpengine_query(sip_msg_t *msg, str *flags)
+{       
+        return rtpengine_rtpp_set_wrap(msg, rtpengine_query_wrap, flags->s, 1);
+}
+
 static int ki_start_recording(sip_msg_t *msg)
 {
 	return rtpengine_rtpp_set_wrap(msg, rtpengine_simple_wrap, NULL, 1, OP_START_RECORDING);
