From 026475aa9d43307442d0e57acf8e11544d00dbed Mon Sep 17 00:00:00 2001
From: Surendra Tiwari <surendratiwari3@gmail.com>
Date: Wed, 9 Jan 2019 08:57:12 +0100
Subject: [PATCH] rtpengine: exported rtpengine_query() variants to KEMI
 framework

- GH #1792
---
 src/modules/rtpengine/rtpengine.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/src/modules/rtpengine/rtpengine.c b/src/modules/rtpengine/rtpengine.c
index 484ce05f3..6269a8c78 100644
--- a/src/modules/rtpengine/rtpengine.c
+++ b/src/modules/rtpengine/rtpengine.c
@@ -3843,6 +3843,16 @@ static int ki_rtpengine_delete(sip_msg_t *msg, str *flags)
 	return rtpengine_rtpp_set_wrap(msg, rtpengine_delete_wrap, flags->s, 1);
 }
 
+static int ki_rtpengine_query0(sip_msg_t *msg)
+{       
+        return rtpengine_rtpp_set_wrap(msg, rtpengine_query_wrap, NULL, 1);
+}
+
+static int ki_rtpengine_query(sip_msg_t *msg, str *flags)
+{       
+        return rtpengine_rtpp_set_wrap(msg, rtpengine_query_wrap, flags->s, 1);
+}
+
 static int ki_start_recording(sip_msg_t *msg)
 {
 	return rtpengine_rtpp_set_wrap(msg, rtpengine_start_recording_wrap, NULL, 1);
@@ -3977,6 +3987,16 @@ static sr_kemi_t sr_kemi_rtpengine_exports[] = {
         { SR_KEMIP_INT, SR_KEMIP_INT, SR_KEMIP_NONE,
             SR_KEMIP_NONE, SR_KEMIP_NONE, SR_KEMIP_NONE }
     },
+    { str_init("rtpengine"), str_init("rtpengine_query0"),
+        SR_KEMIP_INT, ki_rtpengine_query0,
+        { SR_KEMIP_NONE, SR_KEMIP_NONE, SR_KEMIP_NONE,
+            SR_KEMIP_NONE, SR_KEMIP_NONE, SR_KEMIP_NONE }
+    },
+    { str_init("rtpengine"), str_init("rtpengine_query"),
+        SR_KEMIP_INT, ki_rtpengine_query,
+        { SR_KEMIP_STR, SR_KEMIP_NONE, SR_KEMIP_NONE,
+            SR_KEMIP_NONE, SR_KEMIP_NONE, SR_KEMIP_NONE }
+    },
 
     { {0, 0}, {0, 0}, 0, NULL, { 0, 0, 0, 0, 0, 0 } }
 };
-- 
2.20.1

