From: Daniel-Constantin Mierla <miconda@gmail.com>
Date: Fri, 22 Nov 2019 09:42:08 +0100
Subject: [PATCH] systemdops: proper check for null value

---
 src/modules/systemdops/systemdops_mod.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/modules/systemdops/systemdops_mod.c b/src/modules/systemdops/systemdops_mod.c
index 0ac1542..8c24ac1 100644
--- a/src/modules/systemdops/systemdops_mod.c
+++ b/src/modules/systemdops/systemdops_mod.c
@@ -69,7 +69,7 @@ void ksr_sd_app_shutdown(void)
 int mod_register(char *path, int *dlflags, void *p1, void *p2)
 {
 	sr_corecb_t *cbp = sr_corecb_get();
-	if(cbp) {
+	if(cbp==NULL) {
 		return -1;
 	}
 	cbp->app_ready = ksr_sd_app_ready;
