From: S-P Chan <shihping.chan@gmail.com>
Date: Sat, 2 Mar 2024 21:41:11 +0800
Subject: tls: remove unused ENGINE define

---
 src/modules/tls/tls_server.c | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/src/modules/tls/tls_server.c b/src/modules/tls/tls_server.c
index 3e22ec4..3bfea13 100644
--- a/src/modules/tls/tls_server.c
+++ b/src/modules/tls/tls_server.c
@@ -128,11 +128,6 @@ int tls_run_event_routes(struct tcp_connection *c);
 #endif /* __SUNPRO_c */
 #endif /* TLS_RD_DEBUG */
 
-/* only OpenSSL <= 1.1.1 */
-#if !defined(OPENSSL_NO_ENGINE) && OPENSSL_VERSION_NUMBER < 0x030000000L
-#define KSR_SSL_ENGINE
-#endif
-
 extern str sr_tls_xavp_cfg;
 
 static str _ksr_tls_connect_server_id = STR_NULL;
