From: S-P Chan <shihping.chan@gmail.com>
Date: Tue, 13 Feb 2024 17:23:31 +0800
Subject: tls: add logging

---
 src/modules/tls/tls_mod.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/modules/tls/tls_mod.c b/src/modules/tls/tls_mod.c
index 5d3982b..905ca6f 100644
--- a/src/modules/tls/tls_mod.c
+++ b/src/modules/tls/tls_mod.c
@@ -458,6 +458,9 @@ static int mod_child(int rank)
 #else
 	if(rank == PROC_INIT) {
 #endif
+		LM_DBG("Loading SSL_CTX in process_no=%d rank=%d "
+			   "ksr_tls_threads_mode=%d\n",
+				process_no, rank, ksr_tls_threads_mode);
 		if(cfg_get(tls, tls_cfg, config_file).s) {
 			if(tls_fix_domains_cfg(
 					   *tls_domains_cfg, &srv_defaults, &cli_defaults)
