--- a/src/modules/topos/tps_msg.c
+++ b/src/modules/topos/tps_msg.c
@@ -803,6 +803,7 @@ int tps_request_received(sip_msg_t *msg,
 	uint32_t direction = TPS_DIR_DOWNSTREAM;
 	int ret;
 	int use_branch = 0;
+	unsigned int metid = 0;
 
 	LM_DBG("handling incoming request\n");
 
@@ -835,10 +836,11 @@ int tps_request_received(sip_msg_t *msg,
 	if(tps_storage_load_dialog(msg, &mtsd, &stsd) < 0) {
 		goto error;
 	}
-	if(((get_cseq(msg)->method_id) & (METHOD_BYE|METHOD_PRACK|METHOD_UPDATE))
-			&& stsd.b_contact.len <= 0) {
+	metid = get_cseq(msg)->method_id;
+	if((metid & (METHOD_BYE|METHOD_PRACK|METHOD_UPDATE))
+		&& stsd.b_contact.len <= 0) {
 		/* no B-side contact, look for INVITE transaction record */
-		if((get_cseq(msg)->method_id) & (METHOD_UPDATE)) {
+		if(metid & (METHOD_BYE|METHOD_UPDATE)) {
 			/* detect direction - via from-tag */
 			if(tps_dlg_detect_direction(msg, &stsd, &direction) < 0) {
 				goto error;
@@ -916,7 +918,7 @@ int tps_request_received(sip_msg_t *msg,
 				goto error;
 			}
 		}
-		if((get_cseq(msg)->method_id)&(METHOD_SUBSCRIBE)) {
+		if(metid & METHOD_SUBSCRIBE) {
 			if(tps_storage_update_dialog(msg, &mtsd, &stsd, TPS_DBU_CONTACT|TPS_DBU_TIME)<0) {
 				goto error;
 			}
--- a/src/modules/topos/tps_storage.c
+++ b/src/modules/topos/tps_storage.c
@@ -1135,7 +1135,11 @@ int tps_db_load_branch(sip_msg_t *msg, t
 		db_ops[nr_keys]=OP_EQ;
 		db_vals[nr_keys].type = DB1_STR;
 		db_vals[nr_keys].nul = 0;
-		db_vals[nr_keys].val.str_val = TPS_STRZ(md->b_tag);
+		if(md->direction==TPS_DIR_DOWNSTREAM) {
+			db_vals[nr_keys].val.str_val = TPS_STRZ(md->b_tag);
+		} else {
+			db_vals[nr_keys].val.str_val = TPS_STRZ(md->a_tag);
+		}
 		nr_keys++;
 
 		db_keys[nr_keys]=&tt_col_s_method;
