--- a/src/modules/topos/tps_msg.c
+++ b/src/modules/topos/tps_msg.c
@@ -866,7 +866,7 @@ int tps_request_received(sip_msg_t *msg,
 
 	tps_storage_lock_release(&lkey);
 
-	if(use_branch) {
+	if(use_branch && direction == TPS_DIR_DOWNSTREAM) {
 		nuri = stsd.b_contact;
 	} else {
 		if(direction == TPS_DIR_UPSTREAM) {
@@ -884,7 +884,7 @@ int tps_request_received(sip_msg_t *msg,
 		}
 	}
 
-	if(use_branch) {
+	if(use_branch && direction == TPS_DIR_DOWNSTREAM) {
 		if(tps_reappend_route(msg, &stsd, &stsd.s_rr, 1) < 0) {
 			LM_ERR("failed to reappend s-route\n");
 			return -1;
